node/src/components/chain_synchronizer/operations.rs

```diff
@@ -740,156 +793,204 @@ async fn archival_sync(
                 continue;
             }
         };
-        sync_deploys_and_transfers_and_state(effect_builder, &most_recent_block, &node_config)
-            .await?;
+        sync_deploys_and_transfers_and_state(
+            *ctx.effect_builder,
+            &most_recent_block,
+            ctx.node_config,
+        )
+        .await?;
         if let Some(key_block_info) = KeyBlockInfo::maybe_from_block_header(
             most_recent_block.header(),
-            chainspec.protocol_config.verifiable_chunked_hash_activation,
+            ctx.chainspec
+                .protocol_config
+                .verifiable_chunked_hash_activation,
         ) {
             trusted_key_block_info = key_block_info;
         }
     }
     debug!("finish - sync forward - archival sync");
+    debug!("finish - archival_sync - total");
 
     Ok((trusted_key_block_info, most_recent_block.take_header()))
 }
 
 /// Runs the chain synchronization task.
 pub(super) async fn run_chain_sync_task(
     effect_builder: EffectBuilder<JoinerEvent>,
-    trusted_hash: BlockHash,
     chainspec: Arc<Chainspec>,
     node_config: NodeConfig,
+    trusted_hash: BlockHash,
 ) -> Result<BlockHeader, Error> {
     // Fetch the trusted header
-    info!("start - fetch trusted header");
+    debug!("start - fetch trusted header");
     let trusted_block_header = fetch_and_store_block_header(effect_builder, trusted_hash).await?;
     debug!("finish - fetch trusted header");
 
-    if trusted_block_header.protocol_version() > chainspec.protocol_config.version {
+    let chain_sync_context = ChainSyncContext::new(
+        &effect_builder,
+        &chainspec,
+        &node_config,
+        &trusted_block_header,
+        &trusted_hash,
```

Comment:
Isn't `trusted_hash == trusted_block_header.hash()`?

Reply:
It's not about duplication but rather the chance of getting it wrong. If they're equal I think it's just safer to use the hash of the `trusted_block_header`.

Reply:
You're right, I agree. Sorry for the confusion! Let's make it a `trusted_hash` method then.

Reply:
The `trusted_hash()` method has been added here: https://github.com/casper-network/casper-node/pull/2664/commits/a2de4d36785bd2df35a3dda948986a22029facaf